src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+    v = reshape(v, size(alg.φ))
+    v = alg.projector(v, alg.φ)::T
+    v, x
+end
+function KrylovKit.orthogonalize!(v::T, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, x, alg.orth)
+    v = reshape(v, size(alg.φ))
+    v = alg.projector(v, alg.φ)::T
+    v, x
+end
+function KrylovKit.gklrecurrence(operator, U::OrthonormalBasis, V::OrthonormalBasis, β,
+                                 alg::OrthogonalizeAndProject)
+    u = U[end]
+    v = operator(u, true)
+    v = axpy!(-β, V[end], v)
+    # for q in V # not necessary if we definitely reorthogonalize next step and previous step
+    #     v, = orthogonalize!(v, q, ModifiedGramSchmidt())
+    # end
+    α = norm(v)
+    rmul!(v, inv(α))
+
+    r = operator(v, false)
+    r = axpy!(-α, u, r)
+    for q in U
+        r, = orthogonalize!(r, q, alg)
+    end
+    β = norm(r)
+    return v, r, α, β
+end
+
+#
+# Operators
+#
+
+"""
+    apply_Ω(basis::PlaneWaveBasis, δφ, φ, H, λ)
+
+Compute the application of Ω to an element in the space tangent to φ.
+It can be done analitically with the formula
+      Ωδφi = H*δφi - Σj <δφj|H|δφi> δφj - Σj λji δφj
+where λij = <φi|H|φj>;
+"""
+function apply_Ω(basis::PlaneWaveBasis, δφ, φ, H)
+    Nk = length(basis.kpoints)
+    Ωδφ = similar(φ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Hk = H.blocks[ik]
+        λk = φk' * (Hk*φk)
+        Ωδφk = similar(δφk)
+
+        N1 = size(δφk, 2)
+        N2 = size(φk, 2)
+        @assert N1 == N2
+        N = N1
+
+        for i = 1:N
+            Hδφki = Hk * δφk[:, i]
+            Ωδφk[:, i] = Hδφki
+            for j = 1:N
+                Ωδφk[:, i] -= (φk[:, j]'Hδφki) * φk[:, j]
+                Ωδφk[:, i] -= λk[j, i] * δφk[:, j]
+            end
+        end
+        Ωδφ[ik] = Ωδφk
+    end
+    # ensure proper projection onto the tangent space
+    proj_tangent(Ωδφ, φ)
+end
+
+"""
+    apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+
+Compute the application of K to an element in the space tangent to φ
+with the same notations as for the computation of Ω, we have
+      Kδφi = Π(dV * φi),
+Π being the projection on the space tangent to the φ and
+      dV = kernel(δρ)
+where δρ = Σi φi*conj(δφi) + hc.
+"""
+@views function apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+    Nk = length(basis.kpoints)
+    n_spin = basis.model.n_spin_components
+
+    ## compute δρ = Σi φi*conj(δφi)
+    δρ_fourier = zeros(complex(eltype(ρ)), size(ρ)...)
+    for (ik, kpt) in enumerate(basis.kpoints)
+        δρk = zeros(eltype(ρ), basis.fft_size)
+        for i = 1:size(φ[ik], 2)
+            φki_real = G_to_r(basis, kpt, φ[ik][:, i])
+            δφki_real = G_to_r(basis, kpt, δφ[ik][:, i])
+            δρk .+= occ[ik][i] .* (conj.(φki_real) .* δφki_real +
+                                   conj.(δφki_real) .* φki_real)
+        end
+        # Check sanity in the density, we should have ∫δρ = 0
```

Comment:
doesn't seem super useful, it's just checking the orthogonality of the psi, but why not

Reply:
It's to be consistent with compute_density, I like to have it that way.